node/src/db/models/consensus/instructions.rs

```diff
@@ -0,0 +1,275 @@
+                Type::TEXT,
+                Type::TEXT,
+                Type::JSONB,
+            ])
+            .await?;
+
+        let row = client
+            .query_one(&stmt, &[
+                &params.initiating_node_id,
+                &params.signature,
+                &params.asset_id,
+                &params.token_id,
+                &params.template_id,
+                &params.contract_name,
+                &params.status,
+                &params.params,
+                &params.id,
+            ])
+            .await?;
+        Ok(Self::from_row(row)?)
+    }
+
+    /// Marks set of instructions as given status and sets proposal id for reference if provided
+    pub async fn update_instructions_status(
+        instruction_ids: &[InstructionID],
+        proposal_id: Option<ProposalID>,
+        status: InstructionStatus,
+        client: &Client,
+    ) -> Result<(), DBError>
+    {
+        const QUERY: &'static str = "
+            UPDATE instructions SET
+                status = $2,
+                proposal_id = $3,
+                updated_at = NOW()
+            WHERE id::uuid = ANY ($1)";
+        let stmt = client.prepare_typed(QUERY, &[Type::UUID_ARRAY, Type::TEXT]).await?;
+        client
+            .execute(&stmt, &[
+                &instruction_ids.iter().map(|i| i.0).collect::<Vec<uuid::Uuid>>(),
+                &status,
+                &proposal_id,
+            ])
+            .await?;
+
+        Ok(())
+    }
+
+    /// Update instruction state in the database
+    ///
+    /// Updates subset of fields:
+    /// - status
+    /// - proposal_id
+    pub async fn update(self, data: UpdateInstruction, client: &Client) -> Result<Self, DBError> {
+        const QUERY: &'static str = "
+            UPDATE instructions SET
+                status = COALESCE($1, status),
+                proposal_id = $2::\"ProposalID\",
+                updated_at = NOW()
+            WHERE id = $3::\"InstructionID\"
+            RETURNING *";
+        let stmt = client.prepare_typed(QUERY, &[Type::TEXT]).await?;
+        let row = client
+            .query_one(&stmt, &[&data.status, &data.proposal_id, &self.id])
+            .await?;
+        Ok(Self::from_row(row)?)
+    }
+
+    /// Load instruction record
+    pub async fn load(id: InstructionID, client: &Client) -> Result<Self, DBError> {
+        let stmt = "SELECT * FROM instructions WHERE id = $1::\"InstructionID\"";
+        let row = client.query_one(stmt, &[&id]).await?;
+        Ok(Self::from_row(row)?)
+    }
+
+    /// Execute the instruction returning append only state
+    pub async fn execute(
+        &self,
+        _client: &Client,
+    ) -> Result<(Vec<NewAssetStateAppendOnly>, Vec<NewTokenStateAppendOnly>), DBError>
+    {
+        // TODO: we will need to encapsulate the running of an instruction somehow so that nodes can compare view state
+        // to expected state
+        Ok((Vec::new(), Vec::new()))
+    }
```

Comment:
Instruction is a model, we should decopule execute, have InstructionRunner outside of that. InstructionRunner would need to host TemplateContext and pass that along with params to instruction function. I will take it as part of my WIP PR.

Reply:
Actually after more thinking that may work here, in the way that instruction model would collect all the resulting append records from DB. I've published proposed pipeline in the chat, so records might be prepared in parallel as soon as instruction reaches the node. Will rewrite this method as collect_results() and then it would make sense to stay in the model.